torchtune/models/phi4/_tokenizer.py

```diff
@@ -0,0 +1,308 @@
+            trim_leading_whitespace=True,
+        )
+
+        tokenized_messages.extend(encoded)
+        tokenized_messages.append(self.special_tokens["<|im_end|>"])
+        return tokenized_messages
+
+    def tokenize_messages(
+        self,
+        messages: List[Message],
+        *,
+        add_eos: bool = False,
+        ignore_system_prompt: bool = False,
+    ) -> Tuple[List[int], List[bool]]:
+        r"""Tokenize a list of messages one at a time then concatenate them,
+        returning a list of tokens and a list of masks.
+
+        Example:
+            >>> tokenizer = Phi4MiniTokenizer(tokenizer_path, max_seq_len)
+            >>> messages = [
+                Message(role="system", content="system message\n", masked=True),
+                Message(role="user", content="user prompt\n", masked=True),
+                Message(role="assistant", content="assistant response\n"),
+            ]
+
+            >>> # tokenize_messages encodes messages separately and concats
+            >>> tokenizer.tokenize_messages(messages)[0]
+            [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+            >>> # Same result as encoding the full string in one go
+            >>> tokenizer.encode(''.join([message.content for message in messages]))
+            [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+        Args:
+            messages (List[Message]): A list of messages, each containing role, content,
+                and masked attributes.
+            add_eos (bool): Whether to append EOS after assistant message, default to False
+            ignore_system_prompt (bool): Whether to ignore system prompt, defaults to False.
+
+        Raises:
+            ValueError: If the role is not "user", "assistant", or "system".
+
+        Returns:
+            Tuple[List[int], List[bool]]: The tokenized messages
+        """
+        templated_messages = (
+            self.prompt_template(messages)
+            if self.prompt_template is not None
+            else messages
+        )
+
+        start_of_turn = True
+        end_of_turn = False
+        tokenized_messages = []
+        mask = []
+
+        # The chat template in HF adds a bunch of newlines
+        new_line_token_id = self.encode("\n", add_bos=False, add_eos=False)
+
+        for message in templated_messages:
+            # Skip system prompt
+            if ignore_system_prompt and message.role == "system":
+                continue
+
+            # Prepend BOS on start of new turns
+            if start_of_turn:
+                tokenized_messages.append(self.bos_id)
+                mask.append(message.masked)
+
+            if message.role == "assistant":
+                end_of_turn = True
+
+            # Add special tokens
+            tokenized_header = self._tokenize_header(message.role)
+
+            tokenized_messages.extend(tokenized_header)
+
+            # Minus 1, because the len(tokenized_header) is bigger then 1.
```

Comment:
I don't fully follow this, seems like after L241 you just wind up with `mask.extend([message.masked] * len(tokenized_header))` anyways?